src/java.base/share/classes/jdk/internal/misc/X-ScopedMemoryAccess.java.template

```diff
@@ -183,21 +184,56 @@ public class ScopedMemoryAccess {
     }
 
     @ForceInline
-    public void setMemory(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    public void setMemory(AbstractMemorySegmentImpl segment, byte value) {
         try {
-            setMemoryInternal(session, o, offset, bytes, value);
+            setMemoryInternal(segment, segment.byteSize(), value);
         } catch (ScopedAccessError ex) {
             throw ex.newRuntimeException();
         }
     }
 
+    // FILL_NATIVE_THRESHOLD must be a power of two and should be greater than 2^3
+    // Update the value for Aarch64 once 8338975 is fixed.
+    private static final long FILL_NATIVE_THRESHOLD = 1L << (Architecture.isAARCH64() ? 10 : 5);
+
     @ForceInline @Scoped
-    private void setMemoryInternal(MemorySessionImpl session, Object o, long offset, long bytes, byte value) {
+    private void setMemoryInternal(AbstractMemorySegmentImpl segment, long length, byte value) {
+        final MemorySessionImpl session = segment.sessionImpl();
         try {
             if (session != null) {
                 session.checkValidStateRaw();
             }
-            UNSAFE.setMemory(o, offset, bytes, value);
+            // 0...0X...XXXX implies: 0 <= length < FILL_NATIVE_LIMIT
+            if ((length & -FILL_NATIVE_THRESHOLD) == 0) {
+                // Handle smaller segments directly without transitioning to native code
+                final long u = Byte.toUnsignedLong(value);
+                final long longValue = u << 56 | u << 48 | u << 40 | u << 32 | u << 24 | u << 16 | u << 8 | u;
```

Comment:
this can be `u * 0xFFFFFFFFFFFFL` if `value != 0` and just `0L` if not: not sure if fast(er), need to measure